src/fluree/server/components/subscriptions.clj

```diff
@@ -0,0 +1,219 @@
+  (let [{:keys [closed? subs]} @sub-atom]
+    (when-not closed?
+      (doseq [[sub-id {:keys [ws ledgers]}] subs]
+        (when (contains? ledgers ledger-id)
+          ;; TODO - use ledger subscription opts to filter out messages, permission
+          (send-message subscriptions sub-id message))))))
+
+
+(defn send-message-to-all
+  "Sends a message to all subscriptions. Message sent as JSON stringified map
+  in the form: {action: commit, ledger: my/ledger, data: {...}}"
+  [{:keys [sub-atom] :as subscriptions} action ledger-alias data]
+  (log/warn "SEND MESSAGE TO ALL subscriptions: " subscriptions)
+  (let [data*   (if (string? data)
+                  (json/parse data false)
+                  data)
+        message (json/stringify {"action" action
+                                 "ledger" ledger-alias
+                                 "data"   data*})
+        sub-ids (all-sub-ids @sub-atom)]
+    (run! #(send-message subscriptions % message) sub-ids)))
+
+
+(defn close
+  [{:keys [sub-atom] :as _subscriptions}]
+  (log/info "Closing all websocket subscriptions.")
+  (-> sub-atom
+      mark-closed
+      close-all-chans
+      close-all-sockets)
+  (reset! sub-atom {:closed? true}))
+
+(def subscriptions
+  #::ds{:start  (fn [{{:keys [subscription-atom]
+                       :or   {subscription-atom default-subscription-atom}} ::ds/config}]
+                  {:sub-atom subscription-atom})
+        :stop   (fn [{::ds/keys [instance]}]
+                  (close instance))
+        :config {}})
+
+
+(defmulti client-message :msg-type)
+
+(defmethod client-message :on-connect
+  [{:keys [fluree/subscriptions http/sub-id http/sub-chan http/ws]}]
+  (establish-subscription subscriptions ws sub-id sub-chan nil))
+
+(defmethod client-message :on-text
+  [{:keys [fluree/subscriptions http/sub-id payload]}]
+  (try
+    (let [request (json/parse payload true)
+          {:keys [action]} request]
+      (case action
+        "subscribe"
+        (subscribe-ledger subscriptions sub-id request)
+
+        "unsubscribe"
+        (unsubscribe-ledger subscriptions sub-id request)
+
+        "ping" ;; Note in-browser websocket doesn't support ping/pong, so this is a no-op to keep alive
+        (do
+          (log/trace "Websocket keep-alive from sub-id:" sub-id)
+          nil)))
+    (catch Exception e
+      (log/error e "Error with :on-text message from websocket subscriber: " sub-id))))
+
+(defmethod client-message :on-bytes
+  [{:keys [http/sub-id payload offset len]}]
+  (log/info "websocket :on-bytes (no-op) message for sub-id: " sub-id))
+
+;; pong handled automatically, no response needed
+(defmethod client-message :on-ping
+  [{:keys [http/sub-id payload]}]
+  (log/trace "ws :on-ping message received from: " sub-id "with payload: " payload))
+
+(defmethod client-message :on-error
+  [{:keys [http/sub-id error]}]
+  (log/warn "Websocket error for sub-id: " sub-id "with error: " (type error)))
```

Comment:
Should we log this at `error` level? Or are they so common that `warn` is better?

Reply:
Very common, just a network connection break. There is a list of error types, and I'd suspect the optimal solution will be to eventually `case` the error type, with some being log/debug all the way up to log/error. So log/warn is sort of an imperfect compromise. I think we'll have a better sense of what is best here once it is running.